### PR TITLE
Get rid of state in Apache Beam provider hook

### LINKED ISSUE #15536: Get rid of state in Apache Beam provider hook
As discussed in  https://github.com/apache/airflow/pull/15534#discussion_r620500075, we could possibly rewrite Beam Hook to remove the need of storing state in it.

### PR DESCRIPTION
**Pull request by dannikay**
Closes: #15536
Get rid of state in Apache Beam provider hook

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by josh-fell** [comment]
Hi @dannikay! Thanks for contributing! Would you mind providing more context to why this change is desired here please?

**Comment by dannikay** [comment]
Hi @josh-fell https://github.com/apache/airflow/issues/15536 is the context. I remembered that I wrote the issue ID somewhere in the PR (the PR template requires me to fill it in) but I could not find it being displayed anywhere.

**Comment by dannikay** [comment]
Locally triggered "breeze static-checks --all-files --show-diff-on-failure --color always" and it passes with the new patch.

**Comment by dannikay** [comment]
breeze static-checks --all-files --show-diff-on-failure --color always --type run-mypy --all-files
Checking pre-commit installed for /root/.local/pipx/venvs/apache-airflow-breeze/bin/python

Package pre_commit is installed. Good version 3.0.4 (>= 2.0.0)

Good version of Docker: 23.0.1.
Good version of docker-compose: 2.16.0
Good Docker context used: default.
Run mypy for dev.........................................................Passed
Run mypy for core........................................................Passed
Run mypy for providers...................................................Passed
Run mypy for /docs/ folder...............................................Passed

My last run silently failed mypy check due to version mismatch between my local breeze (3.8) and the CI (3.7).

**Comment by potiuk** [comment]
> My last run silently failed mypy check due to version mismatch between my local breeze (3.8) and the CI (3.7).

Mypy always runs with 3.7 (no matter what you use locally to run breeze)

File: airflow/providers/apache/beam/hooks/beam.py [hunk 1]
```diff
@@ -81,81 +81,85 @@ def beam_options_to_args(options: dict) -> list[str]:
...
+
+
+def run_beam_command(
+    cmd: list[str],
+    log: logging.Logger,
+    process_line_callback: Callable[[str], None] | None = None,
+    working_directory: str | None = None,
+) -> None:
     """
     Class responsible for running pipeline command in subprocess
```
Comment (josh-fell) [inline airflow/providers/apache/beam/hooks/beam.py]:
```suggestion
    Function responsible for running pipeline command in subprocess.
```

**Review by josh-fell — APPROVED** [review]
Very cool refactor. One trivial comment update, but overall LGTM.

**Comment by dannikay** [comment]
> Very cool refactor. One trivial comment update, but overall LGTM.

Thanks! Applied the suggestion and all checks passed again.

**Comment by josh-fell** [comment]
@dannikay Woohoo! Great work in your first contribution! 🎉

**Comment by dannikay** [comment]
Thank you for your reviews, @josh-fell !

**Pull request merged**